posts/2022/05/the-evolution-of-the-scipy-developer-cli.md

```diff
@@ -0,0 +1,170 @@
+    def run(cls):
+        print('Running tests...')
+```
+- Command may make use of a Click.Group context defining a `ctx` class attribute
+- Command options are also defined as class attributes
+```python
+@cli.cls_cmd('test')
+class Test():
+    """Run tests"""
+    ctx = CONTEXT
+    verbose = Option(
+        ['--verbose', '-v'], default=False, is_flag=True, help="verbosity")
+    @classmethod
+    def run(cls, **kwargs): # kwargs contains options from class and CONTEXT
+        print('Running tests...')
+```
+
+#### 🎨 `rich-click` addition
+
+To incorporate the look and feel, I designed a layer on top of the existing CLI architecture with the help of [rich-click](https://pypi.org/project/rich-click/).
+It offers a variety of style options, markdown settings and flexibility to group tasks and options. Together it adds that perfect richness to the CLI command pallet.
+Below is a simple example which demonstrates the grouping of options and tasks along with style settings.
+```python
+
+# style and markdown setting
+rich_click.STYLE_ERRORS_SUGGESTION = "yellow italic"
+rich_click.USE_MARKDOWN = True
+
+# grouping global and task based options
+rich_click.OPTION_GROUPS = {
+    "do.py": [
+        {
+            "name": "Options",
+            "options": [
+                "--help", "--build-dir", "--no-build", "--install-prefix"],
+        },
+    ],
+
+    "do.py test": [
+        {
+            "name": "Options",
+            "options": ["--help", "--verbose", "--parallel", "--coverage"],
+        },
+    ],
+}
+# adding tasks into groups
+rich_click.COMMAND_GROUPS = {
+    "do.py": [
+        {
+            "name": "environments",
+            "commands": ["shell", "python", "ipython"],
+        },
+        {
+            "name": "release",
+            "commands": ["notes", "authors"],
+        },
+    ]
+}
+```
+
+#### 🎥 The developer CLI in action
+
+<script id="asciicast-U9l9VvklvEjXdEi1xYS4A7u5M" src="https://asciinema.org/a/U9l9VvklvEjXdEi1xYS4A7u5M.js" async></script>
+
+#### Current list of implemented tasks
+Below are the lists of tasks currently implemented as part of the developer CLI. This list is dynamic and subject to change in the coming months.
+
+* `Build & testing tasks`
+     - build (build & install package on path)
+     - test (Run tests along with options to run tests for a given submodule)
+* `Static checker tasks`
+     - pep8 ( Perform pep8 check with flake8)
+     - mypy ( Run mypy on the codebase)
+* `Environments`
+     - shell (Start Unix shell with PYTHONPATH set)
+     - python (Start a Python shell with PYTHONPATH set)
+     - ipython (Start IPython shell with PYTHONPATH set )
+* `Documentation tasks`
+     - doc (Build documentation)
+     - refguide-check  (Run refguide check)
+
+* `Release tasks`
+    -  notes   (Release notes and log generation)
+    -  authors  (Task to generate list the authors who contributed within a given revision interval)
+* `Benchmarking tasks`
+    - bench & bench compare
+
```

Comment:
excellent! very clear. is the empty line in the bullet list a problem in MD rendering? I don't recall...

Reply:
Markdown I believe. I checked the preview it's rendered properly there